manual/src/refman/extensions/firstclassmodules.etex

```diff
@@ -121,29 +138,89 @@ let draw_using_device device_name picture =
     (val (Hashtbl.find devices device_name) : DEVICE)
   in
   Device.draw picture
+
+(* or, equivalently *)
+let draw_using_device device_name picture =
+  let (module Device : DEVICE) = Hashtbl.find devices device_name in
+  Device.draw picture
 \end{caml_example*}
 
-\lparagraph{p:fst-mod-advexamples}{Advanced examples}
-With first-class modules, it is possible to parametrize some code over the
-implementation of a module without using a functor.
+\lparagraph{p:fst-mod-intro-limitations}{Limitations}
 
-\begin{caml_example}{verbatim}
-let sort (type s) (module Set : Set.S with type elt = s) l =
-  Set.elements (List.fold_right Set.add l Set.empty)
+To avoid breaking type soundness, there is a restriction on functions
+that take module parameters "fun (module A : S) -> ...": if the type
+of the rest of the function (the "...") \emph{depends} on the module
+argument "A", then this function is only allowed to take static module
+arguments of the form "(module M : S)", rather than arbitrary dynamic
+expressions of type "(module S)". We call them \emph{module-dependent}
+functions.
+
+Our "process" function of Paragraph~\ref{ss:fst-mod-intro-mod-fun} is
+such a module-dependent function: its type is "(module M : Monoid) ->
+M.t list -> M.t", where the module "M" is used on the right of the
+arrow "(module M : Monoid) -> ...". The following example is rejected:
+
+\begin{caml_example}{verbatim}[error]
+let sort_order = `Decr
+let wrong =
+  sort
+    (match sort_order with
+     | `Incr -> (module ISetIncr)
+     | `Decr -> (module ISetDecr))
+    [2; 4]
```

Comment:
Maybe also add a version of this example using using `val` to allow keeping a module-dependent function ?
I feel that the presentation here tends to hint that changing the definition of the function `sort` is the way to go but it is not always possible (for example if you didn't define the function or if it uses a parametrized type).